sdk/eventhub/event-hubs/samples/expressSample/src/eventProducer.ts

```diff
@@ -33,32 +30,34 @@ export class EventProducer {
   }
 
   public async send(requestId: string, payload: any) {
-    if (this.batch === undefined) {
-      this.batch = await this.lock.acquire(this.lockKey, () => {
-        if (this.batch === undefined) {
-          console.log(`Acquire lock ${this.lockKey} and create a new batch`);
-          this.lastBatchCreatedTime = Date.now();
-          return this.producerClient.createBatch();
-        } else {
-          console.log(
-            `Acquire lock ${this.lockKey} and skip create. this.batch count: ${this.batch.count}`
-          );
-          return this.batch;
-        }
-      });
-    }
-    const isAdded = this.batch.tryAdd(this.constructEventData(requestId, payload));
+    const batch = await this.getOrCreateBatch();
+    const eventData = this.constructEventData(requestId, payload)
+    const isAdded = batch.tryAdd(eventData);
 
-    if (!isAdded || this.batch.count >= this.batchSendNumber) {
-      const curBatch = this.batch;
-      this.batch = undefined;
-      await this.sendBatch(curBatch);
+    if (!isAdded || batch.count >= this.batchSendNumber) {
+      this.awaitableBatch = undefined;
+      await this.sendBatch(batch);
       if (!isAdded) {
-        console.log(`Up to batch size limit, add event failed. requestId: ${requestId}`);
+        const batch = await this.getOrCreateBatch();
+        if (!batch.tryAdd(eventData)) {
+          console.log(`Up to batch size limit, add event failed. requestId: ${requestId}`);
+        }
       }
     }
   }
 
+  private async getOrCreateBatch(): Promise<EventDataBatch> {
+    // Check if there is an existing promise for a batch.
+    if (this.awaitableBatch) {
+      return this.awaitableBatch;
```

Comment:
It's a great implement! You don't use any await in this function. All codes in this function are executed sequentially. event loop will not switch and avoid risk conditions.  👍

Reply:
No await suggests the `async` keyword could be dropped here for smaller codegen.

Reply:
What does `smaller codegen` mean?